scene/2d/tile_map_layer.cpp

```diff
@@ -2742,6 +2743,10 @@ Vector2i TileMapLayer::local_to_map(const Vector2 &p_pos) const {
 	return tile_set->local_to_map(p_pos);
 }
 
+Vector2i TileMapLayer::collision_to_map(const Vector2 &p_collision_point, const Vector2 &p_collision_normal) const {
+	return local_to_map(to_local(p_collision_point) - p_collision_normal);
```

Comment:
I think normal might be in global coordinates. So it might not work with rotated tilemaps/scenes.
Also, the normal is always of length 1, so I assume if you scale down the TileMap a lot (a tile smaller than one unit), it might "overshoot" the tile. So you might have to normalize the vector in the local coordinates, and maybe multiply its size by 0.5 (as, theoretically, the smallest cell size is 1px wide).

Reply:
I moved the normal into `to_local()` call. I did it at first, but then wrongly though that it doesn't need to be localized.

Reply:
Yeah but that might not be enough if the TileMap is scaled down a lot. 
The problem is that, like, you add the normal to the point in the global space (before transforming it to local space). That means the added normal vector is of length 1px in the global space, and not in the TileMap's one.

To make sure it causes no issue, I think the normal vector's length needs to be adapted to the TileMap's scale. Maybe you can multiply the length of the normal vector by something like `min(tilemap.global_xform.scale.x, tilemap.global_xform.scale.y)` ? (something like that, I didn't try it out).

Reply:
As you mentioned TileMap can also be rotated etc., so direction might change too. This should do what you've suggested I think:
```cpp
Transform2D global_inv = get_global_transform().affine_inverse();
return local_to_map(global_inv.xform(p_collision_point) - 0.5 * global_inv.basis_xform(p_collision_normal).normalized());
```